Add clients exactly

alexkeating opened the pull request:
#### Description

- Migrate clients from the [flexible-voting](https://github.com/ScopeLift/flexible-voting) repo
- There is a corresponding [pr](https://github.com/ScopeLift/flexible-voting/pull/55) in the flexible-voting repo to remove what was migrated.
- No alterations were made a follow up pr will clean up some of the comments and update the flexible-voting dep to match the flexible-voting repo without these client files.

alexkeating committed:
Add clients exactly

alexkeating commented:
Current dependencies on/for this PR:
* main
  * **PR #1** <a href="https://app.graphite.dev/github/pr/ScopeLift/flexible-voting-clients/1" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #2** <a href="https://app.graphite.dev/github/pr/ScopeLift/flexible-voting-clients/2" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/flexible-voting-clients/1?utm_source=stack-comment).

mds1 commented on src/FlexVotingClient.sol:
```diff
@@ -0,0 +1,211 @@
+    //       should still be taken into consideration when determining how much
+    //       weight to vote with this time.
+    // Using the total raw balance to proportion votes in this way means that in
+    // many circumstances this function will not cast votes with all of its
+    // weight.
+    uint256 _totalRawBalanceAtSnapshot = getPastTotalBalance(_proposalSnapshotBlockNumber);
+
+    // We need 256 bits because of the multiplication we're about to do.
+    uint256 _votingWeightAtSnapshot = IVotingToken(address(GOVERNOR.token())).getPastVotes(
+      address(this), _proposalSnapshotBlockNumber
+    );
+
+    //      forVotesRaw          forVoteWeight
+    // --------------------- = ------------------
+    //     totalRawBalance      totalVoteWeight
+    //
+    // forVoteWeight = forVotesRaw * totalVoteWeight / totalRawBalance
+    uint128 _forVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.forVotes) / _totalRawBalanceAtSnapshot
+    );
+    uint128 _againstVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.againstVotes) / _totalRawBalanceAtSnapshot
+    );
+    uint128 _abstainVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.abstainVotes) / _totalRawBalanceAtSnapshot
+    );
+
+    // This param is ignored by the governor when voting with fractional
+    // weights. It makes no difference what vote type this is.
+    uint8 unusedSupportParam = uint8(VoteType.Abstain);
+
+    // Clear the stored votes so that we don't double-cast them.
+    delete proposalVotes[proposalId];
+
+    bytes memory fractionalizedVotes =
+      abi.encodePacked(_againstVotesToCast, _forVotesToCast, _abstainVotesToCast);
+    GOVERNOR.castVoteWithReasonAndParams(
+      proposalId, unusedSupportParam, _castVoteReasonString(), fractionalizedVotes
+    );
+  }
```
Suggestion: dismiss

mds1 commented on src/ATokenFlexVoting.sol:
```diff
@@ -0,0 +1,127 @@
+  /// checkpoint raw balances accordingly.
+  ///
+  /// @inheritdoc MintableIncentivizedERC20
+  function _mint(address account, uint128 amount) internal override {
+    MintableIncentivizedERC20._mint(account, amount);
+    FlexVotingClient._checkpointRawBalanceOf(account);
+    FlexVotingClient.totalBalanceCheckpoints.push(
+      FlexVotingClient.totalBalanceCheckpoints.latest() + amount
+    );
+  }
```
Suggestion: dismiss

mds1 commented on src/FlexVotingClient.sol:
```diff
@@ -0,0 +1,211 @@
+  }
+
+  /// @dev Checkpoints the _user's current raw balance.
+  function _checkpointRawBalanceOf(address _user) internal {
+    balanceCheckpoints[_user].push(_rawBalanceOf(_user));
+  }
```
Suggestion: dismiss

mds1 commented on src/FlexVotingClient.sol:
```diff
@@ -0,0 +1,211 @@
+    //       should still be taken into consideration when determining how much
+    //       weight to vote with this time.
+    // Using the total raw balance to proportion votes in this way means that in
+    // many circumstances this function will not cast votes with all of its
+    // weight.
+    uint256 _totalRawBalanceAtSnapshot = getPastTotalBalance(_proposalSnapshotBlockNumber);
+
+    // We need 256 bits because of the multiplication we're about to do.
+    uint256 _votingWeightAtSnapshot = IVotingToken(address(GOVERNOR.token())).getPastVotes(
+      address(this), _proposalSnapshotBlockNumber
+    );
+
+    //      forVotesRaw          forVoteWeight
+    // --------------------- = ------------------
+    //     totalRawBalance      totalVoteWeight
+    //
+    // forVoteWeight = forVotesRaw * totalVoteWeight / totalRawBalance
+    uint128 _forVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.forVotes) / _totalRawBalanceAtSnapshot
+    );
+    uint128 _againstVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.againstVotes) / _totalRawBalanceAtSnapshot
+    );
+    uint128 _abstainVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.abstainVotes) / _totalRawBalanceAtSnapshot
+    );
+
+    // This param is ignored by the governor when voting with fractional
+    // weights. It makes no difference what vote type this is.
+    uint8 unusedSupportParam = uint8(VoteType.Abstain);
+
+    // Clear the stored votes so that we don't double-cast them.
+    delete proposalVotes[proposalId];
+
+    bytes memory fractionalizedVotes =
+      abi.encodePacked(_againstVotesToCast, _forVotesToCast, _abstainVotesToCast);
+    GOVERNOR.castVoteWithReasonAndParams(
+      proposalId, unusedSupportParam, _castVoteReasonString(), fractionalizedVotes
+    );
+  }
```
Suggestion: fix, return vote count

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/issues/5

mds1 commented on src/ATokenFlexVoting.sol:
```diff
@@ -0,0 +1,127 @@
+  /// checkpoint raw balances accordingly.
+  ///
+  /// @inheritdoc MintableIncentivizedERC20
+  function _burn(address account, uint128 amount) internal override {
+    MintableIncentivizedERC20._burn(account, amount);
+    FlexVotingClient._checkpointRawBalanceOf(account);
+    FlexVotingClient.totalBalanceCheckpoints.push(
+      FlexVotingClient.totalBalanceCheckpoints.latest() - amount
+    );
+  }
```
Suggestion: dismiss

mds1 commented on src/CometFlexVoting.sol:
```diff
@@ -0,0 +1,77 @@
+  //===========================================================================
+  //
+  // This function is called any time the underlying balance is changed.
+  function updateBasePrincipal(address _account, UserBasic memory _userBasic, int104 _principalNew)
+    internal
+    override
+  {
+    Comet.updateBasePrincipal(_account, _userBasic, _principalNew);
+    FlexVotingClient._checkpointRawBalanceOf(_account);
+    FlexVotingClient.totalBalanceCheckpoints.push(uint224(totalSupplyBase));
+  }
```
Suggestion: dismiss

mds1 commented on src/CometFlexVoting.sol:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  /// @notice Returns the current balance in storage for the `account`.
+  function _rawBalanceOf(address account) internal view override returns (uint256) {
+    int104 _principal = userBasic[account].principal;
+    return _principal > 0 ? uint256(int256(_principal)) : 0;
+  }
```
Suggestion: dismiss

mds1 commented on src/FlexVotingClient.sol:
```diff
@@ -0,0 +1,211 @@
+  function _rawBalanceOf(address _user) internal view virtual returns (uint256);
+
+  /// @dev Used as the `reason` param when submitting a vote to `GOVERNOR`.
+  function _castVoteReasonString() internal virtual returns (string memory) {
+    return "rolled-up vote from governance token holders";
+  }
```
Hm, this seems like it should be used when casting votes. So, fix?

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/issues/5

mds1 commented on src/FlexVotingClient.sol:
```diff
@@ -0,0 +1,211 @@
+  /// called multiple times during the lifecycle of a given proposal.
+  /// @param proposalId The ID of the proposal which the Pool will now vote on.
+  function castVote(uint256 proposalId) external {
+    ProposalVote storage _proposalVote = proposalVotes[proposalId];
```
Suggestion: dismiss

mds1 commented:
> No alterations were made a follow up pr will clean up some of the comments and update the flexible-voting dep to match the flexible-voting repo without these client files.

Agreed with not making alterations to code in this PR. As a result, for any of the above code scanning suggestions that we fix, that can be done in a follow up PR

mds1 commented on .github/workflows/ci.yml:
```diff
@@ -22,7 +24,7 @@ jobs:
         run: |
           forge --version
-          forge build --sizes
+          forge build
```
It's disappointing that we have to remove this for *all* clients just because Comet is over the size limit. We should create an issue to track this. Options to add this back are:
1. Get comet below the size limit
2. Delete the comet contracts, then build, then restore them

alexkeating replied:
I am surprised we can't ignore specific contracts in the `forge build` command

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/7cb815322966c9495f5f4a081b4f2a27c53fbf99

mds1 replied:
LGTM, unresolved this thread to add a related comment: Where are we tracking that the Comet contract is above the size limit? My suggestion would be to create an issue for it, and in that issue mention that once it's below the size limit we should undo this change of excluding Comet

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/issues/6

mds1 commented on .github/workflows/ci.yml:
```diff
@@ -75,7 +77,7 @@ jobs:
         with:
           coverage-files: ./lcov.info
-          minimum-coverage: 100 # Set coverage threshold.
+          minimum-coverage: 91 # Set coverage threshold.
```
Why 91, is that just so it passes? If so are we content with keeping it at 91 or should we have an issue to track increasing coverage?

alexkeating replied:
Yes, this is so it passes. I created an issue https://github.com/ScopeLift/flexible-voting-clients/issues/3

mds1 commented on .github/workflows/ci.yml:
```diff
@@ -85,21 +87,24 @@ jobs:
-          scopelint check
+        run: forge fmt --check
+
+      #- name: Install scopelint
```
Why aren't we using scopelint in this repo? If we're permanently not using it we can delete this instead of commenting it out. If we're planning to bring it back let's have an issue to track

alexkeating replied:
Mainly because we can't ignore linting on certain lines. Once this is implemented in scopelint we can comment this back in.

mds1 commented on coverage.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+forge install
```
IMO bash scripts are often hard to read/maintain, especially if you haven't looked at them for a few months. But they are very convenient and the workaround here is nice. However, let's make sure to document what's going on here with code comments—both why this script is needed, and what it's doing / how it works

Also, `forge install` is run automatically if needed, so this line can be removed

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/4dfff53f6f96accfc59fca2b31d115e0b2e161b0

mds1 commented on coverage.sh:
```diff
@@ -0,0 +1,12 @@
+forge install
+find lib/aave-v3-core/contracts -type f -name "*.sol" -exec sed -i 's/pragma solidity 0\.8\.10;/pragma solidity 0.8.13;/g' {} +
+find src/ATokenFlexVoting.sol -type f -name "*.sol" -exec sed -i 's/pragma solidity 0\.8\.10;/pragma solidity 0.8.13;/g' {} +
+find test/ATokenFlexVotingFork.t.sol -type f -name "*.sol" -exec sed -i 's/pragma solidity >=0\.8\.10;/pragma solidity >=0.8.13;/g' {} +
+forge coverage --report summary --report lcov --ir-minimum
+find ./lib/aave-v3-core/contracts -type f -name "*.sol" -exec sed -i 's/pragma solidity 0\.8\.13;/pragma solidity 0.8.10;/g' {} +
```
For example, I'd do something like:

```sh
# First, we change all Aave pragmas from 0.8.10 to 0.8.13, because this is the first solc version with via-ir and we need via-ir because .....
find lib/aave-v3-core/contracts -type f -name "*.sol" -exec sed -i 's/pragma solidity 0\.8\.10;/pragma solidity 0.8.13;/g' {} +
# --- snip --- 

# Now we run coverage
forge coverage --report summary --report lcov --ir-minimum

# Regardless of whether coverage succeeded, we want to change the pragmas back

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/4dfff53f6f96accfc59fca2b31d115e0b2e161b0

mds1 commented on coverage.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+forge install
+find lib/aave-v3-core/contracts -type f -name "*.sol" -exec sed -i 's/pragma solidity 0\.8\.10;/pragma solidity 0.8.13;/g' {} +
```
`sed` behaves differently across platforms, so we might want to use another command here (like `awk`) to ensure it works for all users. For example for me on macOS this coverage script fails:

```
sed: 1: "./lib/aave-v3-core/cont ...": invalid command code .
```

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/4dfff53f6f96accfc59fca2b31d115e0b2e161b0

mds1 commented on script/Deploy.s.sol:
```diff
@@ -1,16 +1,12 @@
-// SPDX-License-Identifier: UNLICENSED
+// SPDX-License-Identifier: MIT
```
There's no executable script in this file, is the plan to delete it or to modify it?

alexkeating replied:
I think scopelint requires this and we plan to switch over.

mds1 replied:
wdym? scopelint doesn't require a script to be present, so we can delete the file if we don't use it

alexkeating replied:
Sorry, what I was referring to is that scopelint requires a script directory. If we delete the file the directory is also [removed](https://github.com/ScopeLift/flexible-voting-clients/tree/261c0a7b3f60bce9c3d74124a1b69ca7428f982d). We could add an empty git.keep file or something so the directory stays.

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/2ae269aa3543c1027e63b1ce17ac9332e4b7addb

mds1 replied:
Ah I see, yea `.gitkeep` makes sense, resolving

mds1 commented on coverage.sh:
```diff
@@ -0,0 +1,21 @@
+find src/ATokenFlexVoting.sol -type f -name "*.sol" -exec awk -i inplace '{gsub(/pragma solidity 0\.8\.10;/, "pragma solidity 0.8.13;"); print}' {} \;
+find test/ATokenFlexVotingFork.t.sol -type f -name "*.sol" -exec awk -i inplace '{gsub(/pragma solidity >=0\.8\.10;/, "pragma solidity >=0.8.13;"); print}' {} \;
+
+# We run the converage with the --ir-minimum flag for the reasons menti --ir-minimum allows for only the stack mover optimizer step which resolving the stack too deep error.
```
> flag for the reasons menti --ir-minimum

typo

alexkeating replied:
https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/1a03faf90d9247784c0f7dbf32839a4020c78df7

mds1 commented on coverage.sh:
```diff
@@ -0,0 +1,21 @@
+
+# First, we change all Aave pragmas from 0.8.10 to 0.8.13, because this is the first solc version with via-ir which we  and we need
+# via-ir because importing the PoolConfigurator contract into the ATokenFlexVoting tests causes a stack too deep error.
+find lib/aave-v3-core/contracts -type f -name "*.sol" -exec awk -i inplace '{gsub(/pragma solidity 0\.8\.10;/, "pragma solidity 0.8.13;"); print}' {} \;
```
I get a ton of warnings like this every time `awk` is invoked. Maybe `awk` has different flags per-platform also?
```
awk: can't open file {gsub(/pragma solidity 0\.8\.10;/, "pragma solidity 0.8.13;"); print}
 source line number 1
awk: unknown option -i ignored

awk: can't open file {gsub(/pragma solidity 0\.8\.10;/, "pragma solidity 0.8.13;"); print}
 source line number 1
awk: unknown option -i ignored
```

I also get this error:

```
[⠢] Compiling...
[⠆] Compiling 93 files with 0.8.10
[⠊] Compiling 60 files with 0.8.15
[⠆] Solc 0.8.10 finished in 17.30s
[⠔] Solc 0.8.15 finished in 21.76s
Error: 
Compiler run failed:
Error: Yul exception:Cannot swap Slot JUNK with Slot RET: too deep in the stack by 1 slots in [ JUNK RET[read_from_storage_split_offset_uint256] _2 expr_26607_component_1_mpos expr_26573_address RET[finalize_allocation] var_ins_mpos _6 _4 var_self_26502_slot _5 _7 _3 _1 _8 var_i expr_11 RET ]
```

alexkeating replied:
Ah okay, I am not seeing those warnings. It must be a cross platform issue. The second error is a result of not having an updated pragma so it should not happen once we get the script working on mac.

alexkeating replied:
Looks like the issue was inplace. Let me know if this works.

https://github.com/ScopeLift/flexible-voting-clients/pull/1/commits/2e58367bc0af4e603ec5999e9dbb7f5c802c02c3

mds1 replied:
Awesome, working as expected now!

```
| File                            | % Lines         | % Statements    | % Branches     | % Funcs        |
|---------------------------------|-----------------|-----------------|----------------|----------------|
| src/ATokenFlexVoting.sol        | 83.33% (10/12)  | 83.33% (10/12)  | 100.00% (0/0)  | 80.00% (4/5)   |
| src/CometFlexVoting.sol         | 100.00% (6/6)   | 100.00% (7/7)   | 100.00% (0/0)  | 100.00% (3/3)  |
| src/FlexVotingClient.sol        | 91.38% (53/58)  | 94.05% (79/84)  | 41.67% (10/24) | 78.57% (11/14) |
| test/ATokenFlexVotingFork.t.sol | 0.00% (0/70)    | 0.00% (0/81)    | 0.00% (0/4)    | 0.00% (0/4)    |
| test/CometFlexVotingFork.t.sol  | 0.00% (0/35)    | 0.00% (0/40)    | 100.00% (0/0)  | 0.00% (0/3)    |
| test/FractionalGovernor.sol     | 0.00% (0/14)    | 0.00% (0/20)    | 100.00% (0/0)  | 0.00% (0/14)   |
| test/GovToken.sol               | 0.00% (0/4)     | 0.00% (0/6)     | 100.00% (0/0)  | 0.00% (0/4)    |
| test/MockATokenFlexVoting.sol   | 100.00% (2/2)   | 100.00% (2/2)   | 100.00% (0/0)  | 100.00% (3/3)  |
| test/ProposalReceiverMock.sol   | 0.00% (0/4)     | 0.00% (0/4)     | 100.00% (0/0)  | 0.00% (0/2)    |
| Total                           | 34.63% (71/205) | 38.28% (98/256) | 35.71% (10/28) | 40.38% (21/52) |
```

apbendi committed:
Remove FlexVotingClient and import it via flexible-voting dependency

apbendi committed:
Fix stack too deep error w/o optimizer by swapping interface for offending Aave contract

Also updates Flexible Voting to v1.1.2 to avoid forge coverage issues related to non-
relative path imports. See release notes at:
https://github.com/ScopeLift/flexible-voting/releases/tag/v1.1.2

apbendi reviewed (approved):
🎉
